File: src/Fable.Ripple/Internal/Tracking.fs
namespace Fable.Ripple.Internal

open Fable.Ripple

/// What is currently being evaluated, and therefore whose reads become
/// dependencies. Owns read tracking, recomputation and edge reconciliation.
module internal Tracking =

    // The node being (re-)evaluated, whose reads become dependencies.
    // `ValueNone` when nothing is tracking.
    let mutable private current: ReactiveNode voption = ValueNone

    // Reads are matched against `current`'s sources in order, so a stable graph
    // reuses its edges instead of rebuilding them.

    // The reads that diverged from the previous sources, collected from the first
    // mismatch on. `ValueNone` while the run still matches - the common case.
    let mutable private currentGets: ResizeArray<ReactiveNode> voption = ValueNone

    // How many leading sources this run has re-read in order; also where the
    // stale tail starts once reconciliation kicks in.
    let mutable private currentGetsIndex = 0

    /// Record that `current` reads `node`. While reads arrive in the same order
    /// as the previous run, just advance the index (no edge mutation); on the
    /// first mismatch, collect the new tail into `currentGets`.
    let track (node: ReactiveNode) =
        match current with
        | ValueNone -> ()
        | ValueSome cur ->
            match currentGets with
            | ValueNone when
                currentGetsIndex < Graph.sourceCount cur
                && obj.ReferenceEquals(Graph.sourceAt cur currentGetsIndex, node)
                ->
                currentGetsIndex <- currentGetsIndex + 1
            | _ ->
                let gets =
                    match currentGets with
                    | ValueSome g -> g
                    | ValueNone ->
                        let g = ResizeArray<ReactiveNode>()
                        currentGets <- ValueSome g
                        g

                gets.Add node

    /// Bring `node` up to date, recomputing only if a dependency truly changed.
    let rec updateIfNecessary (node: ReactiveNode) =
        if node.State = NodeState.Check then
            // A source that actually changed will flip us to Dirty; stop early then.
            let count = Graph.sourceCount node
            let mutable i = 0

            while node.State = NodeState.Check && i < count do
                updateIfNecessary (Graph.sourceAt node i)
                i <- i + 1

        if node.State = NodeState.Dirty then
            recompute node

        node.State <- NodeState.Clean

    and private recompute (node: ReactiveNode) =
        let prevCurrent = current
        let prevGets = currentGets
        let prevIndex = currentGetsIndex
        current <- ValueSome node
        currentGets <- ValueNone
        currentGetsIndex <- 0
        let mutable changed = false

        try
            changed <-
                match node.EffectFn with
                | ValueSome fn ->
                    fn () // effect: run its body; it has no observers to propagate to
                    false
                | ValueNone -> node.Recompute()

            // Reconcile edges only where the dependency set diverged, and only on
            // success - a partial `currentGets` from a throwing evaluation must
            // not touch the graph. A stable graph mutates nothing here.
            match currentGets with
            | ValueSome gets ->
                // Reads diverged at `currentGetsIndex`: drop the old tail, append new.
                Graph.unlinkSourcesTail node currentGetsIndex
                Graph.truncateSources node currentGetsIndex

                for i in 0 .. gets.Count - 1 do
                    let s = gets.[i]
                    Graph.addSource node s
                    Graph.addObserver s node
            | ValueNone ->
                // Read fewer sources than before: drop the stale tail.
                if Graph.sourceCount node > currentGetsIndex then
                    Graph.unlinkSourcesTail node currentGetsIndex
                    Graph.truncateSources node currentGetsIndex
        finally
            // Always restore the tracking globals, even if evaluation threw, so a
            // user exception cannot leave the engine wedged mid-computation. The
            // node stays dirty (its `State <- clean` is skipped) and retries later.
            current <- prevCurrent
            currentGets <- prevGets
            currentGetsIndex <- prevIndex

        if changed then
            Graph.iterObservers
                node
                (fun o ->
                    if int o.State < int NodeState.Dirty then
                        o.State <- NodeState.Dirty
                )

    let untracked (fn: unit -> 'a) =
        let prev = current
        current <- ValueNone

        try
            fn ()
        finally
            current <- prev

File: visuals/src/Fable.Ripple/Internal/Tracking.js

import { addObserver, addSource, truncateSources, unlinkSourcesTail, sourceAt, sourceCount } from "./Graph.js";
import { ReactiveNode__get_RestObservers, ReactiveNode__get_FirstObserver, ReactiveNode__get_Recompute, ReactiveNode__get_EffectFn, ReactiveNode__set_State_Z12CE0414, ReactiveNode__get_State } from "../ReactiveNode.js";
import { item } from "../../../fable_modules/fable-library-js.5.13.0/Array.js";

let current = undefined;

let currentGets = undefined;

let currentGetsIndex = 0;

/**
 * Record that `current` reads `node`. While reads arrive in the same order
 * as the previous run, just advance the index (no edge mutation); on the
 * first mismatch, collect the new tail into `currentGets`.
 */
export function track(node) {
    if (current != null) {
        const cur = current;
        let matchResult;
        if (currentGets == null) {
            if ((currentGetsIndex < sourceCount(cur)) && (sourceAt(cur, currentGetsIndex) === node)) {
                matchResult = 0;
            }
            else {
                matchResult = 1;
            }
        }
        else {
            matchResult = 1;
        }
        switch (matchResult) {
            case 0: {
                currentGetsIndex = ((currentGetsIndex + 1) | 0);
                break;
            }
            case 1: {
                let gets;
                if (currentGets == null) {
                    const g_1 = [];
                    currentGets = g_1;
                    gets = g_1;
                }
                else {
                    gets = currentGets;
                }
                void (gets.push(node));
                break;
            }
        }
    }
}

/**
 * Bring `node` up to date, recomputing only if a dependency truly changed.
 */
export function updateIfNecessary(node) {
    if (ReactiveNode__get_State(node) === 1) {
        const count = sourceCount(node) | 0;
        let i = 0;
        while ((ReactiveNode__get_State(node) === 1) && (i < count)) {
            updateIfNecessary(sourceAt(node, i));
            i = ((i + 1) | 0);
        }
    }
    if (ReactiveNode__get_State(node) === 2) {
        recompute(node);
    }
    ReactiveNode__set_State_Z12CE0414(node, 0);
}

function recompute(node) {
    let matchValue;
    const prevCurrent = current;
    const prevGets = currentGets;
    const prevIndex = currentGetsIndex | 0;
    current = node;
    currentGets = undefined;
    currentGetsIndex = 0;
    let changed = false;
    try {
        changed = ((matchValue = ReactiveNode__get_EffectFn(node), (matchValue == null) ? ReactiveNode__get_Recompute(node)() : ((matchValue(), false))));
        if (currentGets == null) {
            if (sourceCount(node) > currentGetsIndex) {
                unlinkSourcesTail(node, currentGetsIndex);
                truncateSources(node, currentGetsIndex);
            }
        }
        else {
            const gets = currentGets;
            unlinkSourcesTail(node, currentGetsIndex);
            truncateSources(node, currentGetsIndex);
            for (let i = 0; i <= (gets.length - 1); i++) {
                const s = item(i, gets);
                addSource(node, s);
                addObserver(s, node);
            }
        }
    }
    finally {
        current = prevCurrent;
        currentGets = prevGets;
        currentGetsIndex = (prevIndex | 0);
    }
    if (changed) {
        const n = node;
        const option = ReactiveNode__get_FirstObserver(n);
        if (option != null) {
            const o = option;
            if (ReactiveNode__get_State(o) < 2) {
                ReactiveNode__set_State_Z12CE0414(o, 2);
            }
        }
        const option_1 = ReactiveNode__get_RestObservers(n);
        if (option_1 != null) {
            const a = option_1;
            for (let i_1 = 0; i_1 <= (a.length - 1); i_1++) {
                const o = item(i_1, a);
                if (ReactiveNode__get_State(o) < 2) {
                    ReactiveNode__set_State_Z12CE0414(o, 2);
                }
            }
        }
    }
}

export function untracked(fn) {
    const prev = current;
    current = undefined;
    try {
        return fn();
    }
    finally {
        current = prev;
    }
}

